Mathlib/Order/OrderType.lean

```diff
@@ -0,0 +1,450 @@
+    Quotient.inductionOn₃ a b c fun _ _ _ ⟨f⟩ ⟨g⟩ ↦ ⟨f.trans g⟩
+
+theorem type_le {α β}
+    [LinearOrder α] [LinearOrder β] (h : α ↪o β) : type α ≤ type β :=
+  ⟨h⟩
+
+theorem _root_.RelEmbedding.type_le {α β}
+    [LinearOrder α] [LinearOrder β] (h : α ↪o β) : type α ≤ type β :=
+  ⟨h⟩
+
+protected theorem zero_le (o : OrderType) : 0 ≤ o :=
+  inductionOn o (fun _ ↦ OrderEmbedding.ofIsEmpty.type_le)
+
+instance : OrderBot OrderType where
+  bot := 0
+  bot_le := OrderType.zero_le
+
+@[simp]
+theorem bot_eq_zero : (⊥ : OrderType) = 0 :=
+  rfl
+
+instance instIsEmptyIioZero : IsEmpty (Iio (0 : OrderType)) := by
+  simp [← bot_eq_zero]
+
+@[simp]
+protected theorem not_lt_zero (o : OrderType) : ¬o < 0 :=
+  not_lt_bot
+
+instance : ZeroLEOneClass OrderType :=
+  ⟨OrderType.zero_le _⟩
+
+instance instNeZeroOne : NeZero (1 : OrderType) :=
+  ⟨OrderType.one_ne_zero⟩
+
+theorem type_le_iff {α β} [LinearOrder α]
+    [LinearOrder β] : type α ≤ type β ↔ Nonempty (α ↪o β) :=
+  Iff.rfl
+
+theorem type_le_iff' {α β} [inst1 : LinearOrder α]
+    [inst2 : LinearOrder β] : type α ≤ type β ↔ Nonempty (inst1.le ↪r inst2.le) :=
+  ⟨fun f ↦ f, fun f ↦ f⟩
+
+section Cardinal
+/-- The cardinal of an OrderType is the cardinality of any type on which a relation with that order
+type is defined. -/
+def card : OrderType → Cardinal :=
+  Quotient.map _ fun _ _ ⟨e⟩ ↦ ⟨e.toEquiv⟩
+
+@[simp]
+theorem card_type [LinearOrder α] : card (type α) = #α :=
+  rfl
+
+@[gcongr]
+theorem card_le_card {o₁ o₂ : OrderType} : o₁ ≤ o₂ → card o₁ ≤ card o₂ :=
+  inductionOn o₁ fun _ ↦ inductionOn o₂ fun _ _ ⟨f⟩ ↦ ⟨f.toEmbedding⟩
+
+theorem ord_card_mono {o₁ o₂ : OrderType} (h : o₁ ≤ o₂) : (card o₁).ord ≤ (card o₂).ord :=
+  Cardinal.ord_mono (OrderType.card_le_card h)
+
+@[simp]
+theorem card_zero : card 0 = 0 := mk_eq_zero _
+
+@[simp]
+theorem card_one : card 1 = 1 := mk_eq_one _
+
+end Cardinal
+/-- `ω` is the first infinite ordinal, defined as the order type of `ℕ`. -/
+def omega0 : OrderType := type ℕ
+
+/-- The order type of the rational numbers. -/
+def eta : OrderType := type ℚ
+
+/-- The order type of the real numbers on the interval `(0,1)`. -/
+def theta : OrderType := type (Set.Ioo (0 : ℝ) 1)
+
+@[inherit_doc]
+scoped notation "ω" => OrderType.omega0
+
+@[inherit_doc]
+scoped notation "η" => OrderType.eta
+
+@[inherit_doc]
+scoped notation "θ" => OrderType.theta
+
+open Classical
+in instance : Add OrderType.{u} where
+  add := Quotient.map₂ (fun r s ↦ ⟨(r ⊕ₗ s)⟩)
+   (fun _ _ ha _ _ hb ↦ ⟨OrderIso.sumLexCongr (choice ha) (choice hb)⟩)
+
+open Classical
+in instance : HAdd OrderType.{u} OrderType.{v} OrderType.{max u v} where
+  hAdd := Quotient.map₂ (fun r s ↦ ⟨(r ⊕ₗ s)⟩)
+   (fun _ _ ha _ _ hb ↦ ⟨OrderIso.sumLexCongr (choice ha) (choice hb)⟩)
+
+@[simp]
+lemma type_add (α : Type u) (β : Type v) [LinearOrder α] [LinearOrder β] :
+    type (α ⊕ₗ β) = (type α) + (type β) := rfl
+
+lemma OrderIso.sumLexEmpty (α : Type u) [LinearOrder α] : Nonempty (Lex (α ⊕ PEmpty) ≃o α) :=
```

Comment:
a) Why wrap this in `Nonempty`? You can just construct the explicit isomorphism.
b) Can you generalize `PEmpty` to an `IsEmpty` type?
c) Can you put this in a more appropriate file?
d) Same below.